### DISCUSSION TITLE
Add profile for mega cloud backup app

### DISCUSSION THREAD

**Discussion by merrpink** [netblue30/firejail#5508] [Q&A] [answered]
I would like to run mega cloud backup app in firejail. I'm on Fedora37. This app is an rpm and will be installed by dnf. This app is not on Firejail's list so I am not sure how to do this correctly. I want to give this app access to only one directory (the one it will be backing up). I don't want it poking around my system further than that. From what I have been able to understand from the docs, using  `--private=directory` is how I would achieve this. My question is, will Firejail automatically include config files and whatever else the app needs to work correctly or do I need to add these manually? Also, will the app be able to see the files to be backed up in the same directory with it? How do I point to these directories in the app's settings? Is there maybe a better way to set this up?
Thanks

**Comment by ghost** [comment, marked as answer]
Here are some general pointers on creating profiles:
- [creating profiles](https://github.com/netblue30/firejail/wiki/Creating-Profiles)
- [request new profile](https://github.com/netblue30/firejail/issues/1139)

> Will Firejail automatically include config files and whatever else the app needs
or do I need to add these manually?

You will have to add these manually. The workflow goes like this: add blacklists to disable-common.local or disable-programs.local (so other sandboxes can't see them) and noblacklist/whitelist in the app's profile.

> Will the app be able to see the files to be backed up in the same directory with it?

Not sure I understand correctly. Obviously the app will need to be able to see its configuration data, plus the backup dir. Usually these are different locations in the filesystem.

> How do I point to these directories in the app's settings?

When creating a new profile it's common practice to either get the needed info from available documentation and/or running it unsandboxed once to check what directories it creates/accesses so you know what paths are the bare minimum to have a working sandbox.

> Is there maybe a better way to set this up?

In cases like this it can be very insightful to check whether there are existing profiles for applications offering similar functionality. Peeking inside our `dropbox.profile` for example could be a good start.

**Reply by merrpink** [reply]
Thanks for reply. It seems like a complicated and time consuming affair to create a new profile unfortunately. Finding what the app needs from documentation sounds like a major undertaking (probably not published so must check source), and running it unsandboxed to check what it does pretty much defeats the purpose. The reason I am trying to run it in a sandbox is because I don't trust it. If I trust it once, and it is malicious, the damage is already done.

Anyway, I really like the concept of firejail and the list you provide already really is extensive, however I consider most of the apps in the list pretty trustworthy and don't really need to be sandboxed. No easy answers here. Cool app.

**Reply by kmk3** [reply]
@merrpink [on Dec 6](https://github.com/netblue30/firejail/discussions/5508#discussioncomment-4319848):

> Thanks for reply. It seems like a complicated and time consuming affair to
> create a new profile unfortunately. Finding what the app needs from
> documentation sounds like a major undertaking (probably not published so must
> check source)

It can be time consuming, though "what the app needs from documentation"
(assuming you mean the program's documentation) usually boils down to 1-3
directories in the home directory, like the configuration and cache dir, for
example. 

One way to find out what files it uses is to use a mostly empty home directory
with `--private` and see what files it creates in there.  Example:

```sh
firejail --name=foo --private --profile=foo foo

# (run the following in a separate shell)
firejail --join=foo find ~ | LC_ALL=C sort

# And/or open a shell in the sandbox and run commands directly to check what
# files are in there
firejail --join=foo /bin/bash
```

It might be needed to use the program for a while and/or do things like change
its settings to make sure that it creates the files that it uses.

Note: "foo" in `--profile=foo` means `foo.profile`, the profile that you would
create to sandbox the program.

> and running it unsandboxed to check what it does pretty much defeats the
> purpose. The reason I am trying to run it in a sandbox is because I don't
> trust it. If I trust it once, and it is malicious, the damage is already
> done.

Here are two (non-mutually exclusive) ways to mitigate this:

* Run it as sandboxed as possible (see also #5503) and remove commands until it
  works (it may help to look at the program output and `dmesg -H`)
* Run it as a separate user (like Android does for every non-system app), so
  that it at least does not have direct access to your user files

The latter one can be helpful regardless of firejail.

> Anyway, I really like the concept of firejail and the list you provide
> already really is extensive, however I consider most of the apps in the list
> pretty trustworthy and don't really need to be sandboxed. No easy answers
> here. Cool app.

The sandbox restrictions are usually not only targeted towards the program
itself, but also to try to prevent input from untrusted sources from exploiting
a bug in the program, escaping it and trying to run arbitrary code.  For
example, a malicious website trying to exploit a bug in Firefox or a malicious
pdf file doing the same to a program that can parse it.

**Reply by rusty-snake** [reply]
> The reason I am trying to run it in a sandbox is because I don't trust it. If I trust it once, and it is malicious, the damage is already done.

If you can not trust it once because you expect the program itself to be actively harmful, firejail is the wrong tool for you. You should run it on a decidate computer in a strict confined network. My two cents.

**Reply by merrpink** [reply]
@rusty-snake 
You're right. This app is most probably fine and I'm just being a little paranoid, but being a backup tool, I don't want it poking around my filesystem. I tried running it as another user which worked well until I realized the app couldn't write to my backup directory and I couldn't get the directory to maintain group permissions. In addition to chmoding and chowning, I set ACL and umask but it seems that when I use nautilus for copying files, permissions get screwed. 

Firejail looks like it could solve my issues so I guess I'll have to sit down and try to work out the profile...

**Reply by rusty-snake** [reply]
If you only care about messing up `$HOME`, `firejail --private --profile=<foobar> /path/to/program` while developing the profile.

**Reply by merrpink** [reply]
@rusty-snake 
Thanks. `--profile=<foobar>` being the profile I am working on or some other default one?

**Reply by kmk3** [reply]
@merrpink [on Dec 8](https://github.com/netblue30/firejail/discussions/5508#discussioncomment-4339644):

> `--profile=<foobar>` being the profile I am working on or some other default
> one?

The profile you're working on.
